lib/plugins/aws/package/compile/functions.js

```diff
@@ -221,7 +221,7 @@ class AwsCompileFunctions {
 
       functionResource.Properties.Code.S3Key = `${
         this.serverless.service.package.artifactDirectoryName
-      }/${artifactFilePath.split(path.sep).pop()}`;
+      }/${artifactFilePath.split(path.sep).splice(-2).join(path.sep)}`;
```

Comment:
This should properly differentiate for artifacts nested one level, what about if someone uses more deeply nested `artifact`s? I know it's quite exotic example, but I was wondering, is there any danger in using the whole provided artifact path ?

Reply:
I've quickly tested this example:

```
functions:
  hello:
    handler: my-function.handler
    package:
      artifact: other/nested/artifact/testservicename.zip

  hello2:
    handler: my-function.handler
    package:
      artifact: very/nested/artifact/testservicename.zip
```

And the S3Keys are identical for both functions. 

@medikoo do you think it's fine to support one-level nesting or we should support all possible nestings?

Reply:
@pgrzesik That was the solution suggested in https://github.com/serverless/serverless/issues/4128#issue-251476320, but the implementation might be indeed flawed, or at least won't support deeper nesting levels.
I'll think about more universal approach in the meantime.

Reply:
I believe we should support all possible nesting.

I think right solution is to simply strip `<service-path>/.serverless` from a full path. Still we need to also confirm how it works with custom artifacts (they're not put into `.serverless`). Maybe right logic is to:
- If file is in `<service-path>/.serverless/` strip `<service-path>/.serverless/`
- Otherwise strip `<service-path>` (and throw if artifact is not even in `<service-path>`, that I believe should not be supported

Reply:
Thanks for explanation @scadu :bow: 

@medikoo for custom artifacts I believe the path is exactly the same as defined in `serverles.yml` so in the case of the example I used, path is equal to `other/nested/artifact/testservicename.zip` In such cases we can just use it as-is I believe

Reply:
@pgrzesik I tested your example on the current master and my fork and the S3Key is not identical in this case:
```
❯ Select-String "S3Key" .\.serverless\serverless-state.json

.serverless\serverless-state.json:182:                "S3Key": "serverless/zo
mgservice/dev/1612552187918-2021-02-05T19:09:47.918Z/other/nested/artifact/ar
tifact.zip"
.serverless\serverless-state.json:207:                "S3Key": "serverless/zo
mgservice/dev/1612552187918-2021-02-05T19:09:47.918Z/very/nested/artifact/art
ifact.zip"
```

`serverless.yml`:
```yml
service: zomgservice

configValidationMode: error
frameworkVersion: '*'

provider:
  name: aws

functions:
  hello:
    handler: my-function.handler
    package:
      artifact: other/nested/artifact/artifact.zip

  hello2:
    handler: my-function.handler
    package:
      artifact: very/nested/artifact/artifact.zip
```
So, at this moment I'm not even able to reproduce the issue. However, I'll continue with different scenarios to make sure there's no clash and try with @medikoo's suggestions.

I marked the PR as draft for now.

Reply:
That's interesting - however if I'm not mistaken, you're on Windows, right? If that's the case, `other/nested/artifact/artifact.zip` will be treated as a single item when splitted on `path.sep`. I've just tested again and got identical keys. Let me know if that's the case on your side :bow:

Reply:
@pgrzesik Indeed, it behaves differently on Windows and Linux. The current state of the PR should behave the same on Windows and Linux. I didn't introduce any conditionals mentioned above.

Reply:
Thanks @scadu - I've tested that approach locally and while it's working perfectly for explicitly defined custom artifacts as in my example, it's breaking for functions that use artifacts built with the Framework

For this definition:
```
hello3:
    handler: my-function.handler
    package:
      individually: true
```

That's the `S3Key` that is being generated on my machine:
```
serverless/test--servicename/dev/1612786076339-2021-02-08T12:07:56.339Z//home/pgrzesik/Resources/work-serverless/reproduce/.serverless/hello3.zip
```

which is invalid. I believe the best way would be to take @medikoo suggestion to first strip everything up to `.serverless` directory in path. Please let me know what do you think

Reply:
No worries, it's a tricky one :100: